3_core_core_analysis/7_stability_vs_genome_location.py

```diff
@@ -0,0 +1,384 @@
+
+        pa14_loc = pa14_core_genes_df.loc[mapped_pa14_id, "PA14_scaled"]
+
+        # Calculate the difference
+        pao1_vs_pa14_loc = abs(pao1_loc - pa14_loc)
+
+        diff_lst.append(pao1_vs_pa14_loc)
+
+    return diff_lst
+
+
+pao1_least_dist = dist_pao1_pa14_homolog(
+    pao1_similarity, pa14_similarity, gene_mapping_pao1, window, pao1_least_stable
+)
+
+pao1_most_dist = dist_pao1_pa14_homolog(
+    pao1_similarity, pa14_similarity, gene_mapping_pao1, window, pao1_most_stable
+)
+
+pao1_random_dist = dist_pao1_pa14_homolog(
+    pao1_similarity, pa14_similarity, gene_mapping_pao1, window, random_gene_set
+)
+
+# ## Plot
+
+# +
+# Format data for plotting
+homolog_neighborhood_df = pd.DataFrame(
+    data={
+        "least stable % matching": least_matched_neighborhood,
+        "most stable % matching": most_matched_neighborhood,
+        "random % matching": random_matched_neighborhood,
+    }
+)
+
+homolog_neighborhood_df = (
+    homolog_neighborhood_df.stack()
+    .reset_index()
+    .rename(columns={"level_1": "core gene group", 0: "value"})
+)
+
+# +
+homolog_dist_df = pd.DataFrame(
+    data={
+        "least stable dist": pao1_least_dist,
+        "most stable dist": pao1_most_dist,
+        "random dist": pao1_random_dist,
+    }
+)
+
+homolog_dist_df = (
+    homolog_dist_df.stack()
+    .reset_index()
+    .rename(columns={"level_1": "core gene group", 0: "value"})
+)
+
+# +
+plt.figure(figsize=(10, 8))
+fig_match_homolog = sns.stripplot(
+    data=homolog_neighborhood_df, x="core gene group", y="value", palette="Set2"
+)
+
+plt.title("% Matching homologs", fontsize=14)
+plt.xlabel("")
+plt.xticks(fontsize=14)
+plt.ylabel("% neighboring homologs that match", fontsize=14)
+
+# +
+plt.figure(figsize=(10, 8))
+fig_dist = sns.stripplot(
+    data=homolog_dist_df, x="core gene group", y="value", palette="Set2"
+)
+
+plt.title("Distance between homologs", fontsize=14)
+plt.xlabel("")
+plt.xticks(fontsize=14)
+plt.ylabel("L1 Distance", fontsize=14)
+
+# +
+# Save
+fig_match_homolog.figure.savefig(
+    "stability_percent_match_homolog.svg",
+    format="svg",
+    bbox_inches="tight",
+    transparent=True,
+    pad_inches=0,
+    dpi=300,
+)
+
+fig_dist.figure.savefig(
+    "stability_dist.svg",
+    format="svg",
+    bbox_inches="tight",
+    transparent=True,
+    pad_inches=0,
+    dpi=300,
+)
+# -
+
+# Both approaches yeild similar trends which is good. So the signal is robust. However, both approaches find that there is not an association between stability and if the gene is located in the same location across strain types. We thought that least stable genes might be located in a different location in PAO1 vs PA14 (i.e. we would expect L1 distance to be higher in least stable genes or percent matching homologs to be lower in least stable genes). We instead find that the distances and percent of matching homologs are similar between most and least stable genes.
```

Comment:
Not a dumb question. I'll try to clarify the comments. Basically the goal of this analysis was to determine if least stable genes are located in the same location on the genome in PAO1 and PA14. There were two approaches we used to answer this:

Approach 1: starts with the least stable core gene in PAO1 and selects its nearest core neighbors. Then map that least stable core gene to PA14. Starting with the homologous least stable core gene in PA14 select its nearest core neighbors. Next, since these neighbors are core genes can map between PAO1 and PA14. We compare the two sets of neighbors and determine how many of the neighbors in PAO1 map to the PA14 neighbors. If the percent of neighbors that map are high then this indicates that the starting least stable gene is located a similar location in both PAO1 and PA14 because the neighboring genes are the same.

Approach 2: Takes the genome location defined by the gene ids and scales them to be in the same range since the number of genes in PA14 is more than in PAO1. Then we take the absolute value of the difference between the least stable gene in PAO1 versus PA14. 

So both approaches are trying to answer the same question, we cannot directly calculate distance because the strains have different genome lengths so we have two approximations